<a href="https://colab.research.google.com/github/YaninaK/anomaly-detection/blob/b1/notebooks/02_Anomalies_1_2_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Обнаружение аномалий в начислениях за тепловую энергию


## 2. Аномалии 1, 2 и 4.



### Примеры аномалий

Виды аномалий по показаниям приборов учёта тепловой энергии, которые необходимо выявлять (кроме объектов с видом энергопотребления ГВС (централ):

1. нулевые значения показаний за тепловую энергию в отопительный период (октябрь-апрель);

2. равные значения показаний в течение нескольких расчетных периодов;

3. снижение/рост показаний в отдельные месяцы по сравнению с показаниями за предыдущие периоды по данному объекту (с учётом фактической температуры наружного воздуха и количества отопительных дней в месяце);

4. аномально низкое/высокое (отклонение более 25%) потребление объекта в конкретном месяце по сравнению с аналогичными объектами (только для типов объекта «Многоквартирный дом») по критериям:
  - год постройки (по группам до 1958 г., 1959-1989 гг., 1990-2000 гг., 2001-2010 гг., 2011-2024 гг.),
  - этажность (по группам 1-2 этажа, 3-4 этажа, 5-9 этажей,10-12 этажей, 13 и более этажей),
  - площадь (±10%),
  - наличие ГВС ИТП (горячей воды, учитываемой тем же прибором).



In [1]:
initiate = True
if initiate:
  !git init -q
  !git clone -b b1  https://github.com/YaninaK/anomaly-detection.git -q

  from google.colab import drive
  drive.mount('/content/drive')

  !unzip -u -q /content/drive/MyDrive/ML_projects/08_anomaly_detection/data/01_raw/task#3.zip -d /content/anomaly-detection/data/01_raw

%cd /content/anomaly-detection/notebooks

Mounted at /content/drive
/content/anomaly-detection/notebooks


In [2]:
import os
import sys

sys.path.append(os.getcwd())
sys.path.append(os.path.join(os.getcwd(), "..", "src", "anomaly_detection"))

In [3]:
import datetime

import numpy as np
import pandas as pd

from data.make_dataset import load_data
from data.preprocess import Preprocess
from features.data_sequence import generate_data_sequence
from features.missing_records import (
    select_missing_records, select_uninvoiced_objects, select_nonunique_objects
)
from features.duplicated import get_equal_values
from features.grouping import Grouping
from features.period_outliers import get_outlers

In [4]:
import warnings
warnings.filterwarnings('ignore')

In [5]:
PATH = "/content/anomaly-detection/"

## 1. Чтение данных

In [6]:
folder_path = '../data/01_raw/'

In [7]:
regenerate = True
data, temperature, buildings = load_data(folder_path, regenerate, path=PATH)

100%|██████████| 24/24 [00:14<00:00,  1.62it/s]


## 2. Подготовка данных

In [8]:
preprocess = Preprocess()
data, buildings = preprocess.fit_transform(data, buildings)

data.shape

(75385, 10)

In [9]:
df = generate_data_sequence(data)
df.shape

print(f"df.shape = {df.shape}\n")
df.sample(2)

df.shape = (4628, 24)



2021-07-01  \
Адрес объекта                     Тип объекта         № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                 
г Уфа, ул. Первомайская, д.26     Многоквартирный дом 106655 1               г Уфа, ул. Первомайская, д.26          11.254   
г Уфа, пр-кт Октября, д.81 корп.1 Многоквартирный дом 244979 1               г Уфа, пр-кт Октября, д.81 корп.1       7.952   

                                                                                                                2021-08-01  \
Адрес объекта                     Тип объекта         № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                 
г Уфа, ул. Первомайская, д.26     Многоквартирный дом 106655 1               г Уфа, ул. Первомайская, д.26          11.254   
г Уфа, пр-кт Октября, д.81 корп.1 Многоквартирный дом 244979 1               г Уфа, пр-кт Октября, д.81 корп.1       8.855   

                                                                                                                2021-09-01  \
Адрес объекта                     Тип объекта         № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                 
г Уфа, ул. Первомайская, д.26     Многоквартирный дом 106655 1               г Уфа, ул. Первомайская, д.26          22.706   
г Уфа, пр-кт Октября, д.81 корп.1 Многоквартирный дом 244979 1               г Уфа, пр-кт Октября, д.81 корп.1       9.185   

                                                                                                                2021-10-01  \
Адрес объекта                     Тип объекта         № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                 
г Уфа, ул. Первомайская, д.26     Многоквартирный дом 106655 1               г Уфа, ул. Первомайская, д.26          93.664   
г Уфа, пр-кт Октября, д.81 корп.1 Многоквартирный дом 244979 1               г Уфа, пр-кт Октября, д.81 корп.1      49.756   

                                                                                                                2021-11-01  \
Адрес объекта                     Тип объекта         № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                 
г Уфа, ул. Первомайская, д.26     Многоквартирный дом 106655 1               г Уфа, ул. Первомайская, д.26         126.807   
г Уфа, пр-кт Октября, д.81 корп.1 Многоквартирный дом 244979 1               г Уфа, пр-кт Октября, д.81 корп.1      64.193   

                                                                                                                2021-12-01  \
Адрес объекта                     Тип объекта         № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                 
г Уфа, ул. Первомайская, д.26     Многоквартирный дом 106655 1               г Уфа, ул. Первомайская, д.26         153.094   
г Уфа, пр-кт Октября, д.81 корп.1 Многоквартирный дом 244979 1               г Уфа, пр-кт Октября, д.81 корп.1      78.195   

                                                                                                                2022-01-01  \
Адрес объекта                     Тип объекта         № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                 
г Уфа, ул. Первомайская, д.26     Многоквартирный дом 106655 1               г Уфа, ул. Первомайская, д.26         193.598   
г Уфа, пр-кт Октября, д.81 корп.1 Многоквартирный дом 244979 1               г Уфа, пр-кт Октября, д.81 корп.1      96.053   

                                                                                                                2022-02-01  \
Адрес объекта                     Тип объекта         № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                 
г Уфа, ул. Первомайская, д.26     Многоквартирный дом 106655 1               г Уфа, ул. Первомайская, д.26         150.598   
г Уфа, пр-кт Октября, д.81 корп.1 Многоквартирный дом 244979 1               г Уфа, пр-кт Октября, д.81 корп.1      71.619   

                                                                        

## 3. Аномалии



### 3.1 Нулевые значения показаний за тепловую энергию в отопительный период (октябрь-апрель)

#### 3.1.1 Пропуски в данных по текущему потреблению

In [10]:
save=True
all_periods = False
missing_records = select_missing_records(df, all_periods, save, path=PATH)

print(f"missing_records.shape = {missing_records.shape}\n")
missing_records.sample(2)

missing_records.shape = (1053, 14)



2021-10-01 00:00:00  \
Адрес объекта                      Тип объекта     № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                           
г Уфа, ул. Кирова, д.41 корп.43    Другое строение 440550 0               г Уфа, ул. Кирова, д.41 корп.43                     NaN   
г Уфа, ул. Российская, д.11 корп.2 Другое строение 97599  0               г Уфа, ул. Российская, д.11 корп.2               25.301   

                                                                                                              2021-11-01 00:00:00  \
Адрес объекта                      Тип объекта     № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                           
г Уфа, ул. Кирова, д.41 корп.43    Другое строение 440550 0               г Уфа, ул. Кирова, д.41 корп.43                     NaN   
г Уфа, ул. Российская, д.11 корп.2 Другое строение 97599  0               г Уфа, ул. Российская, д.11 корп.2               38.378   

                                                                                                              2021-12-01 00:00:00  \
Адрес объекта                      Тип объекта     № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                           
г Уфа, ул. Кирова, д.41 корп.43    Другое строение 440550 0               г Уфа, ул. Кирова, д.41 корп.43                  11.370   
г Уфа, ул. Российская, д.11 корп.2 Другое строение 97599  0               г Уфа, ул. Российская, д.11 корп.2               46.416   

                                                                                                              2022-01-01 00:00:00  \
Адрес объекта                      Тип объекта     № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                           
г Уфа, ул. Кирова, д.41 корп.43    Другое строение 440550 0               г Уфа, ул. Кирова, д.41 корп.43                     NaN   
г Уфа, ул. Российская, д.11 корп.2 Другое строение 97599  0               г Уфа, ул. Российская, д.11 корп.2               60.939   

                                                                                                              2022-02-01 00:00:00  \
Адрес объекта                      Тип объекта     № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                           
г Уфа, ул. Кирова, д.41 корп.43    Другое строение 440550 0               г Уфа, ул. Кирова, д.41 корп.43                     NaN   
г Уфа, ул. Российская, д.11 корп.2 Другое строение 97599  0               г Уфа, ул. Российская, д.11 корп.2                  NaN   

                                                                                                              2022-03-01 00:00:00  \
Адрес объекта                      Тип объекта     № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                           
г Уфа, ул. Кирова, д.41 корп.43    Другое строение 440550 0               г Уфа, ул. Кирова, д.41 корп.43                     NaN   
г Уфа, ул. Российская, д.11 корп.2 Другое строение 97599  0               г Уфа, ул. Российская, д.11 корп.2                41.88   

                                                                                                              2022-04-01 00:00:00  \
Адрес объекта                      Тип объекта     № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                           
г Уфа, ул. Кирова, д.41 корп.43    Другое строение 440550 0               г Уфа, ул. Кирова, д.41 корп.43                     NaN   
г Уфа, ул. Российская, д.11 корп.2 Другое строение 97599  0               г Уфа, ул. Российская, д.11 корп.2               27.758   

                                                                                                              2022-10-01 00:00:00  \
Адрес объекта                      Тип объекта     № ОДПУ Вид энерг-а ГВС Адрес объекта 2                                           
г Уфа, ул. Кирова, д.41 корп.43    Другое строение 440550 0               г Уфа, ул. Кирова, д.41 корп.43        

#### 3.1.2 Объекты без данных по текущему потреблению

In [11]:
save = True

uninvoiced_objects = select_uninvoiced_objects(df, buildings, save, path=PATH)

print(f"uninvoiced_objects.shape = {uninvoiced_objects.shape}")
uninvoiced_objects.sample(2)

uninvoiced_objects.shape = (1650, 6)


Адрес объекта          Тип Объекта  \
275   г Уфа, ул. Богдана Хмельницкого, д.45  Многоквартирный дом   
1186                г Уфа, ул. Попова, д.17          Частный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  \
275                   2     1961-01-01                  563.6   
1186                  1     1951-01-01                   56.5   

                            Адрес объекта 2  
275   г Уфа, ул. Богдана Хмельницкого, д.45  
1186                г Уфа, ул. Попова, д.17

In [12]:
pd.concat(
    [
        uninvoiced_objects["Тип Объекта"].value_counts(),
        uninvoiced_objects["Тип Объекта"].value_counts(normalize=True)
    ], axis=1
).head(10)

count  proportion
Тип Объекта                                           
Многоквартирный дом                    752    0.455758
Другое строение                        652    0.395152
Частный дом                            138    0.083636
Административные здания, конторы        45    0.027273
Учебное заведение, комбинат, центр      13    0.007879
Школы и ВУЗ                             13    0.007879
Гаражи                                   6    0.003636
Жилое здание (Гостиница, Общежитие)      5    0.003030
Производственный объект                  4    0.002424
Магазины                                 3    0.001818

In [13]:
cond = uninvoiced_objects["Тип Объекта"] == "Многоквартирный дом"
s = uninvoiced_objects[cond]['Общая площадь объекта'].sum() / 1e06

print(f"У {cond.sum()} объектов типа Многоквартирный дом общей площадью{s: 0.1f} млн. нет данных о выствленных счетах.")

У 752 объектов типа Многоквартирный дом общей площадью 1.1 млн. нет данных о выствленных счетах.


* У 1650 объектов нет данных об учете потребления теплоэнергии. Из них 752 объекта (46%) общей площадью 1.1 млн. приходится на тип ```Многоквартирный дом```, около 40% - на тип ```Другое строение```, около 8.3% - на тип ```Частный дом```.


#### 3.1.3 Неуникальные адреса объектов.

In [14]:
save = True
nonunique = select_nonunique_objects(buildings, save, path=PATH)

print(f"Всего неуникальных объектов: {nonunique.shape[0]}\n")
nonunique.head(6)

Всего неуникальных объектов: 145



Адрес объекта      Тип Объекта  Этажность объекта  \
122  г Уфа, б-р. Тухвата Янаби, д.34  Другое строение                  1   
123  г Уфа, б-р. Тухвата Янаби, д.34  Другое строение                  2   
412      г Уфа, пр-кт Октября, д.153  Другое строение                  0   
413      г Уфа, пр-кт Октября, д.153  Другое строение                  3   
487       г Уфа, пр-кт Октября, д.31  Другое строение                  0   
488       г Уфа, пр-кт Октября, д.31  Другое строение                  5   

    Дата постройки  Общая площадь объекта                  Адрес объекта 2  
122            NaT                   0.01  г Уфа, б-р. Тухвата Янаби, д.34  
123            NaT                   0.01  г Уфа, б-р. Тухвата Янаби, д.34  
412            NaT                   0.01      г Уфа, пр-кт Октября, д.153  
413            NaT                   0.01      г Уфа, пр-кт Октября, д.153  
487     1970-01-01                   0.01       г Уфа, пр-кт Октября, д.31  
488     1970-01-01                   0.01       г Уфа, пр-кт Октября, д.31

In [15]:
pd.concat(
    [
      nonunique["Тип Объекта"].value_counts(),
      nonunique["Тип Объекта"].value_counts(True),
      nonunique.groupby(["Тип Объекта"])["Адрес объекта"].agg(lambda x: x.nunique())
    ], axis=1
)

count  proportion  Адрес объекта
Тип Объекта                                                         
Другое строение                       116    0.800000             51
Многоквартирный дом                    14    0.096552              7
Частный дом                            12    0.082759              4
Учебное заведение, комбинат, центр      3    0.020690              1

1. Типы объектов ```Другое строение``` и ```Частный дом``` оставлены без корректировок:
  * Уникальный адрес объекта нужен для привязки площади объекта и даты постройки к данным учета потребления теплоэнергии. У типа ```Другое строение``` - в большинстве случаев эти данные отстутствуют, соответственно этот блок информации не имеет смысла корректировать.

  * Информации об неуникальных объектах типа ```Частный дом``` нет в данных о потреблении теплоэнергии - этот блок также не имеет смысла корректировать.

2. Блок ```Многоквартирный дом``` скорректирован:
  * Адреса многоквартирных домов сделаны уникальными в признаке ```Адрес объекта 2``` через сопоставление с данными об учете теплоэнергии.
  * На трех объектах ниже счета не выставляются, для них в признаке ```Адрес объекта 2``` к адресу добавлено слово ```extra``` :
    * ```г Уфа, ул. Вологодская, д.20```,
    * ```г Уфа, ул. Интернациональная, д.113```,
    * ```г Уфа, ул. Нежинская, д.6```.
3. Блок ```Учебное заведение, комбинат, центр``` скорректирован:
  * Адреса сделаны уникальными в признаке ```Адрес объекта 2``` через сопоставление с данными об учете теплоэнергии и проставление ``№ ОДПУ`` в адрес.

In [16]:
nonunique_2 = buildings[
    buildings.duplicated(subset=["Адрес объекта 2", "Тип Объекта"], keep=False)
]
pd.concat(
    [
      nonunique_2["Тип Объекта"].value_counts(),
      nonunique_2["Тип Объекта"].value_counts(True),
      nonunique_2.groupby(["Тип Объекта"])["Адрес объекта 2"].agg(lambda x: x.nunique())
    ], axis=1
)

count  proportion  Адрес объекта 2
Тип Объекта                                        
Другое строение    115    0.905512               51
Частный дом         12    0.094488                4

* Если использовать ```Адрес объекта 2```, неуникальные адреса в типах  ```Многоквартирный дом``` и ```Учебное заведение, комбинат, центр``` появляться не будут.

## 3.2 Равные значения показаний в течение нескольких расчетных периодов

In [17]:
save = True
equal_values = get_equal_values(data, save, path=PATH)

print(f"Число записей с равными значениями показаний: {equal_values.shape[0]}")
print(f"Число адресов с равными значениями показаний: {equal_values['Адрес объекта'].nunique()}\n")
equal_values.tail(10)

Число записей с равными значениями показаний: 242
Число адресов с равными значениями показаний: 94



index Подразделение                    № ОДПУ Вид энерг-а ГВС  \
65236   4145           Уфа                   1524327            None   
57777   2746           Уфа                   1524327            None   
19082   3738           Уфа  00535101751, 00533101751            None   
46913   4292           Уфа  00535101751, 00533101751            None   
13060   4294           Уфа                  65896908            None   
42496   4285           Уфа                  65896908            None   
24894   4069           Уфа                    101433            None   
54942   4231           Уфа                    101433            None   
36090   3524           Уфа                      1981            None   
57663   2632           Уфа                      1981            None   

                                         Адрес объекта          Тип объекта  \
65236  г Уфа, ул. Энтузиастов, д.14, Подобъект №984513  Многоквартирный дом   
57777  г Уфа, ул. Энтузиастов, д.14, Подобъект №984513  Многоквартирный дом   
19082  г Уфа, ул. Энтузиастов, д.16, Подобъект №984098  Многоквартирный дом   
46913  г Уфа, ул. Энтузиастов, д.16, Подобъект №984098  Многоквартирный дом   
13060  г Уфа, ул. Энтузиастов, д.16, Подобъект №984816  Многоквартирный дом   
42496  г Уфа, ул. Энтузиастов, д.16, Подобъект №984816  Многоквартирный дом   
24894                   г Уфа, ул. Юрия Гагарина, д.20      Другое строение   
54942                   г Уфа, ул. Юрия Гагарина, д.20      Другое строение   
36090                        р-н Уфимский, д Лекаревка      Другое строение   
57663                        р-н Уфимский, д Лекаревка      Другое строение   

      Дата текущего показания  Текущее потребление, Гкал Период потребления  
65236              2022-03-01                      1.000         2022-02-01  
57777              2022-04-01                      1.000         2022-03-01  
19082              2021-12-01                      2.500         2021-11-01  
46913              2022-12-01                      2.500         2022-11-01  
13060              2023-04-01                      2.519         2023-03-01  
42496              2023-05-01                      2.519         2023-04-01  
24894              2021-11-01                     11.009         2021-10-01  
54942              2022-11-01                     11.009         2022-10-01  
36090              2022-01-01                    210.000         2021-12-01  
57663              2022-04-01                    210.000         2022-03-01

## 3.3. Аномально низкое/высокое (отклонение более 25%) потребление объекта в конкретном месяце по сравнению с аналогичными объектами

(только для типов объекта «Многоквартирный дом») по критериям:
  - год постройки (по группам до 1958 г., 1959-1989 гг., 1990-2000 гг., 2001-2010 гг., 2011-2024 гг.),
  - этажность (по группам 1-2 этажа, 3-4 этажа, 5-9 этажей,10-12 этажей, 13 и более этажей),
  - площадь (±10%),
  - наличие ГВС ИТП (горячей воды, учитываемой тем же прибором).

In [18]:
grouping = Grouping()
df_grouped, dropped = grouping.fit_transform(df, buildings)

print(f"df_grouped.shape = {df_grouped.shape}\n")
df_grouped.sample(2)

df_grouped.shape = (3204, 35)



Адрес объекта          Тип Объекта  \
409   г Уфа, пр-кт Октября, д.65 корп.1  Многоквартирный дом   
3009    г Уфа, ул. Чудинова, д.1 корп.1  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  \
409                 5.0     1968-01-01                 4501.2   
3009                9.0     1972-01-01                 3882.5   

                        Адрес объекта 2 Группа этажность объекта  \
409   г Уфа, пр-кт Октября, д.65 корп.1               5-9 этажей   
3009    г Уфа, ул. Чудинова, д.1 корп.1               5-9 этажей   

               Улица Дата постройки 2 Группа год постройки  ...  \
409    пр-кт Октября       1968-01-01        1959-1989 гг.  ...   
3009    ул. Чудинова       1972-01-01        1959-1989 гг.  ...   

      2022-09-01 00:00:00  2022-10-01 00:00:00  2022-11-01 00:00:00  \
409                13.917               63.977               99.859   
3009                  NaN               72.363               85.734   

      2022-12-01 00:00:00  2023-01-01 00:00:00  2023-02-01 00:00:00  \
409               137.181              152.902              135.227   
3009              116.757              121.254              105.641   

      2023-03-01 00:00:00  2023-04-01 00:00:00  2023-05-01 00:00:00  \
409               110.756              101.814               12.398   
3009               83.344               50.323                  NaN   

      2023-06-01 00:00:00  
409                14.241  
3009                  NaN  

[2 rows x 35 columns]

In [19]:
dropped

Адрес объекта          Тип Объекта  Этажность объекта  \
1432  г Уфа, ул. Кирова, д.95  Многоквартирный дом                1.0   

     Дата постройки  Общая площадь объекта  \
1432     1971-01-01                   0.01   

                                Адрес объекта 2 Группа этажность объекта  \
1432  г Уфа, ул. Кирова, д.95, Подобъект №46590                1-2 этажа   

            Улица Дата постройки 2 Группа год постройки  ...  \
1432   ул. Кирова       1971-01-01        1959-1989 гг.  ...   

      2022-09-01 00:00:00  2022-10-01 00:00:00  2022-11-01 00:00:00  \
1432                  NaN                6.623               18.004   

      2022-12-01 00:00:00  2023-01-01 00:00:00  2023-02-01 00:00:00  \
1432               14.127               29.448               25.116   

      2023-03-01 00:00:00  2023-04-01 00:00:00  2023-05-01 00:00:00  \
1432               18.209               11.972                  NaN   

      2023-06-01 00:00:00  
1432                  NaN  

[1 rows x 35 columns]

Удалены объекты с неуказанной общей площадью:

1. Нет информации ```Этажность объекта```, ```Дата постройки```, ```Общая площадь объекта``` по многоквартирному дому по адресу ```г Уфа, ул. Уфимское шоссе, д.4```.

2. По адресу ```г Уфа, ул. Кирова, д.95``` значатся 2 многоквартирных дома: 5 этажей и 1 этаж. По одноэтажному дому (```Подобъект №46590```) нет информации по общей площади объекта.

In [20]:
df_grouped.groupby(
    ["Группа год постройки", "Группа этажность объекта", "Вид энерг-а ГВС"]
)["Общая площадь объекта"].count()

Группа год постройки  Группа этажность объекта  Вид энерг-а ГВС
до 1958 г             1-2 этажа                 0                     7
                                                1                     3
                      3-4 этажа                 0                    92
                                                1                    79
                      5-9 этажей                0                    79
                                                1                    34
                      10-12 этажей              0                     1
                                                1                     0
                      13 и более этажей         0                     0
                                                1                     2
1959-1989 гг.         1-2 этажа                 0                    11
                                                1                     0
                      3-4 этажа                 0                    35
                                                1                    20
                      5-9 этажей                0                  1367
                                                1                   406
                      10-12 этажей              0                   168
                                                1                    11
                      13 и более этажей         0                    14
                                                1                     4
1990-2000 гг.         1-2 этажа                 0                     0
                                                1                     0
                      3-4 этажа                 0                     1
                                                1                     1
                      5-9 этажей                0                   205
                                                1                    34
                      10-12 этажей              0                   100
                                                1                    19
                      13 и более этажей         0                    20
                                                1                    12
2001-2010 гг.         1-2 этажа                 0                     0
                                                1                     0
                      3-4 этажа                 0                     1
                                                1                     0
                      5-9 этажей                0                    48
                                                1                    21
                      10-12 этажей              0                    61
                                                1                    34
                      13 и более этажей         0                    56
                                                1                    45
2011-2024 гг.         1-2 этажа                 0                     0
                                                1                     0
                      3-4 этажа                 0                     0
                                                1                     0
                      5-9 этажей                0                     1
                                                1                     6
                      10-12 этажей              0                     5
                                                1                    26
                      13 и более этажей         0                    28
                                                1                   146
Name: Общая площадь объекта, dtype: int64

* Наибольшее число объектов в группе
  * год постройки: 1959-1989 гг.
  * этажность объекта: 5-9 этажей
  * Вид энерг-а ГВС: 0

In [21]:
save=True
n_periods = 24
threshold = 0.25

outliers_addr, consumption_mask = get_outlers(
    df_grouped, threshold=threshold, n_periods=n_periods,  save=save, path=PATH
)
print(f"Aномально низкое/высокое (отклонение более {threshold:.0%}) потребление:")
for period in outliers_addr:
  n_underconsumption_addr = len(outliers_addr[period]["underconsumption"])
  n_overconsumption_addr = len(outliers_addr[period]["overconsumption"])
  print(f"{period: %Y-%m}:  {n_underconsumption_addr}\t{n_overconsumption_addr}")

Aномально низкое/высокое (отклонение более 25%) потребление:
 2021-07:  154	202
 2021-08:  182	193
 2021-09:  207	253
 2021-10:  361	579
 2021-11:  211	358
 2021-12:  207	281
 2022-01:  213	268
 2022-02:  194	274
 2022-03:  221	285
 2022-04:  286	388
 2022-05:  151	193
 2022-06:  135	192
 2022-07:  153	202
 2022-08:  175	213
 2022-09:  133	169
 2022-10:  475	770
 2022-11:  217	339
 2022-12:  205	289
 2023-01:  206	272
 2023-02:  182	277
 2023-03:  220	296
 2023-04:  371	473
 2023-05:  147	204
 2023-06:  168	225


In [22]:
print(f"consumption_mask.shape = {consumption_mask.shape}\n")
consumption_mask.sample(2)

consumption_mask.shape = (3204, 35)



Адрес объекта          Тип Объекта  \
124   г Уфа, б-р. Хадии Давлетшиной, д.17  Многоквартирный дом   
2701              г Уфа, ул. Свободы, д.6  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  \
124                10.0     1979-01-01                 4060.9   
2701                5.0     1964-01-01                 3749.7   

                          Адрес объекта 2 Группа этажность объекта  \
124   г Уфа, б-р. Хадии Давлетшиной, д.17             10-12 этажей   
2701              г Уфа, ул. Свободы, д.6               5-9 этажей   

                        Улица Дата постройки 2 Группа год постройки  ...  \
124    б-р. Хадии Давлетшиной       1979-01-01        1959-1989 гг.  ...   
2701              ул. Свободы       1964-01-01        1959-1989 гг.  ...   

      2022-09-01 00:00:00  2022-10-01 00:00:00  2022-11-01 00:00:00  \
124                   NaN                  3.0                  2.0   
2701                  NaN                  2.0                  2.0   

      2022-12-01 00:00:00  2023-01-01 00:00:00  2023-02-01 00:00:00  \
124                   2.0                  2.0                  2.0   
2701                  2.0                  2.0                  2.0   

      2023-03-01 00:00:00  2023-04-01 00:00:00  2023-05-01 00:00:00  \
124                   2.0                  2.0                  NaN   
2701                  2.0                  2.0                  NaN   

      2023-06-01 00:00:00  
124                   NaN  
2701                  NaN  

[2 rows x 35 columns]

* Объекты, на которых встречалось аномально низкое потребление - на 25% ниже медианного в рамках группы:
  - год постройки (по группам до 1958 г., 1959-1989 гг., 1990-2000 гг., 2001-2010 гг., 2011-2024 гг.),
  - этажность (по группам 1-2 этажа, 3-4 этажа, 5-9 этажей,10-12 этажей, 13 и более этажей),
  - площадь (±10%),
  - наличие ГВС ИТП (горячей воды, учитываемой тем же прибором).
* В ```consumption_mask``` помечены цифрой 1.

In [23]:
underconsumption = df_grouped[(consumption_mask.iloc[:, -n_periods:] == 1).sum(axis=1) > 0]

print(f"underconsumption.shape = {underconsumption.shape}\n")
underconsumption.sample(2)

underconsumption.shape = (1274, 35)



Адрес объекта          Тип Объекта  Этажность объекта  \
2620  г Уфа, ул. Российская, д.50  Многоквартирный дом                9.0   
1708     г Уфа, ул. Ленина, д.162  Многоквартирный дом               10.0   

     Дата постройки  Общая площадь объекта              Адрес объекта 2  \
2620     1984-01-01                11062.3  г Уфа, ул. Российская, д.50   
1708     1996-01-01                10451.0     г Уфа, ул. Ленина, д.162   

     Группа этажность объекта            Улица Дата постройки 2  \
2620               5-9 этажей   ул. Российская       1984-01-01   
1708             10-12 этажей       ул. Ленина       1996-01-01   

     Группа год постройки  ...  2022-09-01 00:00:00  2022-10-01 00:00:00  \
2620        1959-1989 гг.  ...                  NaN                6.123   
1708        1990-2000 гг.  ...              38.5467              154.320   

      2022-11-01 00:00:00  2022-12-01 00:00:00  2023-01-01 00:00:00  \
2620               44.079              53.0270               69.149   
1708              289.660             374.8091              402.382   

      2023-02-01 00:00:00  2023-03-01 00:00:00  2023-04-01 00:00:00  \
2620              59.8550               45.004               31.631   
1708             320.3489              236.901              164.218   

      2023-05-01 00:00:00  2023-06-01 00:00:00  
2620                  NaN                  NaN  
1708              48.6174              39.4989  

[2 rows x 35 columns]

* Объекты, на которых встречалось аномально высокое потребление - на 25% выше медианного в рамках группы:
  - год постройки (по группам до 1958 г., 1959-1989 гг., 1990-2000 гг., 2001-2010 гг., 2011-2024 гг.),
  - этажность (по группам 1-2 этажа, 3-4 этажа, 5-9 этажей,10-12 этажей, 13 и более этажей),
  - площадь (±10%),
  - наличие ГВС ИТП (горячей воды, учитываемой тем же прибором).
* В ```consumption_mask``` помечены цифрой 3.

In [24]:
overconsumption = df_grouped[(consumption_mask.iloc[:, -n_periods:] == 3).sum(axis=1) > 0]

print(f"overconsumption.shape = {overconsumption.shape}\n")
overconsumption.sample(2)

overconsumption.shape = (1662, 35)



Адрес объекта          Тип Объекта  \
2033  г Уфа, ул. Набережная р. Уфы, д.11/2А  Многоквартирный дом   
2358           г Уфа, ул. Пожарского, д.261  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  \
2033               10.0     2013-01-01                 5822.3   
2358                9.0     1992-01-01                 4095.3   

                            Адрес объекта 2 Группа этажность объекта  \
2033  г Уфа, ул. Набережная р. Уфы, д.11/2А             10-12 этажей   
2358           г Уфа, ул. Пожарского, д.261               5-9 этажей   

                       Улица Дата постройки 2 Группа год постройки  ...  \
2033   ул. Набережная р. Уфы       2013-01-01        2011-2024 гг.  ...   
2358          ул. Пожарского       1992-01-01        1990-2000 гг.  ...   

      2022-09-01 00:00:00  2022-10-01 00:00:00  2022-11-01 00:00:00  \
2033               27.162               68.407              112.473   
2358                  NaN               62.568               96.357   

      2022-12-01 00:00:00  2023-01-01 00:00:00  2023-02-01 00:00:00  \
2033              154.387              171.416              140.524   
2358              133.655              140.540              117.479   

      2023-03-01 00:00:00  2023-04-01 00:00:00  2023-05-01 00:00:00  \
2033              119.065               92.631               36.054   
2358               89.075               77.701                  NaN   

      2023-06-01 00:00:00  
2033               15.142  
2358                  NaN  

[2 rows x 35 columns]

* Объекты с потреблением +/- 25% от медианного потребление в рамках группы (описаны выше).
* В ```consumption_mask``` помечены цифрой 2.

In [25]:
normal = df_grouped[(consumption_mask.iloc[:, -n_periods:] == 2).sum(axis=1) > 0]

print(f"normal consumption.shape = {normal.shape}\n")
normal.sample(2)

normal consumption.shape = (3014, 35)



Адрес объекта          Тип Объекта  \
2436             г Уфа, ул. Рихарда Зорге, д.23  Многоквартирный дом   
2767  г Уфа, ул. Степана Халтурина, д.53 корп.1  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  \
2436                5.0     1967-01-01                 3189.9   
2767                5.0     1974-01-01                 4365.5   

                                Адрес объекта 2 Группа этажность объекта  \
2436             г Уфа, ул. Рихарда Зорге, д.23               5-9 этажей   
2767  г Уфа, ул. Степана Халтурина, д.53 корп.1               5-9 этажей   

                       Улица Дата постройки 2 Группа год постройки  ...  \
2436       ул. Рихарда Зорге       1967-01-01        1959-1989 гг.  ...   
2767   ул. Степана Халтурина       1974-01-01        1959-1989 гг.  ...   

      2022-09-01 00:00:00  2022-10-01 00:00:00  2022-11-01 00:00:00  \
2436                  NaN               34.494               73.104   
2767                  NaN               69.888              102.069   

      2022-12-01 00:00:00  2023-01-01 00:00:00  2023-02-01 00:00:00  \
2436             111.2930              113.372              95.6940   
2767             150.0908              152.026             127.6641   

      2023-03-01 00:00:00  2023-04-01 00:00:00  2023-05-01 00:00:00  \
2436               76.500               52.073                  NaN   
2767              100.014               83.333                  NaN   

      2023-06-01 00:00:00  
2436                  NaN  
2767                  NaN  

[2 rows x 35 columns]